src/lib.rs

```diff
@@ -29,40 +30,69 @@ pub enum Algorithm {
 /// A part of the JWT: header and claims specifically
 /// Allows converting from/to struct with base64
 pub trait Part {
-    fn from_base64(encoded: String) -> Result<Self, Error> where Self: Sized;
-    fn to_base64(&self) -> Result<String, Error>;
+    type Encoded: AsRef<str>;
+
+    fn from_base64<B: AsRef<[u8]>>(encoded: B) -> Result<Self, Error> where Self: Sized;
+    fn to_base64(&self) -> Result<Self::Encoded, Error>;
 }
 
 impl<T> Part for T where T: Encodable + Decodable {
-    fn to_base64(&self) -> Result<String, Error> {
+    type Encoded = String;
+
+    fn to_base64(&self) -> Result<Self::Encoded, Error> {
         let encoded = try!(json::encode(&self));
         Ok(encoded.as_bytes().to_base64(base64::URL_SAFE))
     }
 
-    fn from_base64(encoded: String) -> Result<T, Error> {
-        let decoded = try!(encoded.as_bytes().from_base64());
+    fn from_base64<B: AsRef<[u8]>>(encoded: B) -> Result<T, Error> {
+        let decoded = try!(encoded.as_ref().from_base64());
         let s = try!(String::from_utf8(decoded));
         Ok(try!(json::decode(&s)))
     }
 }
 
-#[derive(Debug, PartialEq, RustcEncodable, RustcDecodable)]
+#[derive(Debug, PartialEq)]
 /// A basic JWT header part, the alg is automatically filled for use
 /// It's missing things like the kid but that's for later
 pub struct Header {
-    typ: String,
+    typ: &'static str,
     alg: Algorithm,
 }
 
 impl Header {
     pub fn new(algorithm: Algorithm) -> Header {
         Header {
-            typ: "JWT".to_owned(),
+            typ: "JWT",
             alg: algorithm,
         }
     }
 }
 
+impl Part for Header {
+    type Encoded = &'static str;
+
+    fn from_base64<B: AsRef<[u8]>>(encoded: B) -> Result<Self, Error> where Self: Sized {
+        let algoritm = match encoded.as_ref() {
+            b"eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9" => { Algorithm::HS256 },
+            b"eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzM4NCJ9" => { Algorithm::HS384 },
+            b"eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9" => { Algorithm::HS512 },
+            _ => return Err(Error::InvalidToken)
+        };
+
+        Ok(Header::new(algoritm))
+    }
+
+    fn to_base64(&self) -> Result<Self::Encoded, Error> {
+        let encoded = match self.alg {
+            Algorithm::HS256 => { "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9" },
+            Algorithm::HS384 => { "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzM4NCJ9" },
+            Algorithm::HS512 => { "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9" },
+        };
+
+        Ok(encoded)
+    }
+}
```

Comment:
I think it can work in tandem, if `kid` or something dynamic is used then do the computation, else use the "known" values. (the `Encoded` assoc-type could be [`Cow<'static, str>`](http://doc.rust-lang.org/std/borrow/enum.Cow.html))